storage/remote/queue_manager.go

```diff
@@ -443,11 +443,21 @@ func (t *QueueManager) AppendMetadata(ctx context.Context, metadata []scrape.Met
 		})
 	}
 
-	err := t.sendMetadataWithBackoff(ctx, mm)
-
-	if err != nil {
-		t.metrics.failedMetadataTotal.Add(float64(len(metadata)))
-		level.Error(t.logger).Log("msg", "non-recoverable error while sending metadata", "count", len(metadata), "err", err)
+	var i int
```

Comment:
> keep track of the start and end indexes into the slice for deciding what samples to send

I'm not sure what you mean here.

Also, `math.Ceil` isn't needed because the decimals are implicitly truncated when assigning to an `int` (and we don't want to round, must truncate).

Reply:
I'm saying we should maintain the decimal places (cast one of the two ints to a float) so that we can round up if `len(metadata)` is not evenly divisible by `MaxSamplesPerSend`. Either that or you could use `len(metadata) % MaxSamplesPerSend` and increment `i` if the remainder was not 0.

> keep track of the start and end indexes into the slice for deciding what samples to send

I meant you'd need to a way to track what indexes within the metadata slice are a part of the next metadata send, just so that on the last send we index `slice[beginning of last send:end of slice]` instead of `slice[beginning of last send:beginning plus MaxSamplesPerSend]`

Either way, this would get rid of the need for the if condition at the end that checks if we need to do one last send.

Reply:
Hows this:

```golang
for i = 0; i < int(math.Ceil(float64(len(metadata))/float64(t.mcfg.MaxSamplesPerSend)); i++ {
                last := (i+1)*t.mcfg.MaxSamplesPerSend
                if last > len(metadata) {
                        last = len(metadata)
                }
		err := t.sendMetadataWithBackoff(ctx, mm[i*t.mcfg.MaxSamplesPerSend:last])
		if err != nil {
			t.metrics.failedMetadataTotal.Add(float64(t.mcfg.MaxSamplesPerSend))
			level.Error(t.logger).Log("msg", "non-recoverable error while sending metadata", "count", t.mcfg.MaxSamplesPerSend, "err", err)
		}
}

Reply:
Personally I would put the  `math.Ceil` calculation on it's own line with something like `numSends := math.Ceil(...)` but this looks good :+1: